src/function/accumulated.rs

```diff
@@ -69,15 +69,30 @@ where
             // output vector, we want to push in execution order, so reverse order to
             // ensure the first child that was executed will be the first child popped
             // from the stack.
-            let origin = zalsa
+            let Some(origin) = zalsa
                 .lookup_ingredient(k.ingredient_index)
-                .origin(db, k.key_index);
-            let inputs = origin.iter().flat_map(|origin| origin.inputs());
+                .origin(db, k.key_index)
+            else {
+                continue;
+            };
+
+            let estimated_inputs = match &origin {
+                QueryOrigin::Assigned(_) | QueryOrigin::BaseInput => 0,
+                QueryOrigin::Derived(edges) | QueryOrigin::DerivedUntracked(edges) => {
+                    edges.input_outputs.len()
+                }
+            };
+
+            stack.reserve(estimated_inputs);
```

Comment:
Might as well just
```suggestion
            if let QueryOrigin::Derived(edges) | QueryOrigin::DerivedUntracked(edges) = &origin {
                    stack.reserve(edges.input_outputs.len());
            }
```
which looks a bit simpler imo

Reply:
That's great. Thanks